NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,531 @@
+        '''
+        # create an empty list for a placeholder to store all the texts
+        all_files = []
+        
+        # search for text files (.txt) inside the folder and extract all the texts
+        for input_file in txt_upload.value.keys():
+            text_dict = {}
+            
+            # use the text file name as the doc_id
+            doc_id = input_file.replace('.txt', '')
+            
+            try:
+                # read the text file
+                doc_lines = codecs.decode(txt_upload.value[input_file]['content'], encoding='utf-8')
+                
+                # store them inside a dictionary
+                text_dict['text_id'] = doc_id
+                text_dict['text'] = doc_lines
+                all_files.append(text_dict)
+                    
+            except:
+                # this will provide some information in the case of an error
+                app_logger.exception("message")
+                traceback.print_exc()
+        
+        # convert the extracted texts into a pandas dataframe for further processing
+        text_df = pd.DataFrame.from_dict(all_files)
+        text_df['spacy_text'] = text_df['text'].apply(lambda text: self.nlp_preprocess(text))
+        text_df.set_index('text_id', inplace=True)
+        
+        return text_df
+    
+    
+    def process_xls(self, xls_upload):
+        '''
+        Pre-process uploaded .xlsx file into pandas dataframe
+
+        Args:
+            xls_upload: the uploaded .xlsx file from upload_files()
+        '''
+        # read the excel file containing the list of texts and convert them into a pandas dataframe
+        text_df = pd.read_excel(io.BytesIO(xls_upload.data[0]))
+        text_df['spacy_text'] = text_df['text'].apply(lambda text: self.nlp_preprocess(text))
+        text_df.set_index('text_id', inplace=True)
+        
+        return text_df
+        
+
+    def get_quotes(self, text_df, inc_ent, create_tree=False):
+        '''
+        Extract quotes and their meta-data (quote_id, quote_index, etc.) from the text
+        and return as a pandas dataframe
+
+        Args:
+            text_df: the pandas dataframe containing the list of texts
+            inc_ent: a list containing the named entities to be extracted from the text, 
+                     e.g., ['ORG','PERSON','GPE','NORP','FAC','LOC']
+            create_tree: option to create parse tree files for the quotes 
+        '''
+        # create an output folder and specify the file path if create_tree=True
+        if create_tree:
+            os.makedirs('output', exist_ok=True)
+            os.makedirs('output/trees', exist_ok=True)
+            tree_dir = './output/trees/'
+        else:
+            tree_dir = None
+    
+        # create an empty list to store all detected quotes
+        all_quotes = []
+        
+        # go through all the texts and start extracting quotes
+        for row in text_df.itertuples():
+            doc_id = row.Index
```

Comment:
Wondering if we can generate the doc_id/text_id other than sequential. Every time we upload a set of files, it's always starting from 0. We probably should use the hash of the file content as the text_id, so that these are relatively unique and consistent, and irrelevant to the order of uploading?

Reply:
there are ways to generate hash from the content (https://stackoverflow.com/questions/42819622/getting-hash-digest-of-a-file-in-python-reading-whole-file-at-once-vs-readin), but the unique id is less relevant compared to file names. I've made changes to the code to use the first 10 char of the file names + add numbering for duplicates - let me know which one you'd prefer

Reply:
Displaying (trucated) filenames in the file selection box (we should use combobox https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#Combobox for its wildcard search function) is good, for dataframe slicing, using unique (and fixed length) ID like hash would be a good option. First 10 char of the filenames sometimes won't reserve enough identifiable information for the user to choose from.

Reply:
![Screen Shot 2022-07-07 at 2 42 08 pm](https://user-images.githubusercontent.com/19829189/177693031-e2f2ed32-2cb3-49e0-9eb4-bd9cb27761c9.png)
For example, I uploaded a bunch of files just now, and the file selector is shown like this.